### PR TITLE
Config gem and Add method to verify apple user identity

### PR DESCRIPTION
**Pull request by TimoPeraza**
#### Description:
* Do some configutation to the gem: gemfile.spec, moving version gem, adding dependencies, adding codeclimata
* Add a class to offer a method which permit valid the user identity in base of the json web token
* Add tests
---

### CODE REVIEW

File: apple_sign_in.gemspec [hunk 1]
```diff
@@ -1,34 +1,41 @@
...
-  spec.authors       = ['fedeagripa']
-  spec.email         = ['federicogagripa@gmail.com']
+  spec.version       = AppleSignIn::Base::VERSION
+  spec.authors       = ['Timothy Peraza']
+  spec.email         = ['timothy@rootstrap.com']
 
-  spec.summary       = ''
-  spec.description   = ''
-  spec.homepage      = ''
+  spec.summary       = 'A gem that implement an integration with apple sign in'
```
Comment (mrubi-rootstrap) [inline apple_sign_in.gemspec:11]: How about describing the gem with `Integration with Sign In with Apple authentication`?

File: .gitignore [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 # rspec failure tracking
 .rspec_status
+.byebug_history
+.DS_Store
```
Comment (mrubi-rootstrap) [inline .gitignore:13]: Yes, thanks!

File: lib/apple_sign_in/base.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+require 'apple_sign_in/base/version'
+
+module AppleSignIn
+  module Base
+    class Error < StandardError; end
+    # Your code goes here...
```
Comment (mrubi-rootstrap) [inline lib/apple_sign_in/base.rb:8]: The comment can be deleted

File: lib/apple_sign_in/user_identity.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    def initialize(user_identify, jwt)
+      @user_identify = user_identify
+      @jwt = jwt
+    end
+
+    def valid?
+      token_data = apple_token_from_jwk
+
+      return true if token_data.key?('sub') && user_identify == token_data['sub']
```
Comment (mrubi-rootstrap) [inline lib/apple_sign_in/user_identity.rb:17]: The 'sub' short name can be a bit cryptic for those not familiarised with jwt. Maybe a refactor extracting the logical expression to

```ruby
def match_user_identity?
  token_data.key?('sub') && user_identify == token_data['sub']
end
```

could be more intention revealing

File: lib/apple_sign_in/user_identity.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: false
+
+module AppleSignIn
+  class UserIdentity
+    APPLE_KEY_URL = 'https://appleid.apple.com/auth/keys'.freeze
+
+    attr_reader :user_identify, :jwt
+
+    def initialize(user_identify, jwt)
+      @user_identify = user_identify
```
Comment (mrubi-rootstrap) [inline lib/apple_sign_in/user_identity.rb:10]: I think there's a typo in `user_identify` instead of `user_identity`

**Comment by mrubi-rootstrap** [comment]
Awesome, thanks! I left a few comments

File: spec/user_identity_spec.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  subject(:user_identity_service) { described_class.new(uid, signed_jwt) }
+
+  context '#valid?' do
+    context 'when the parameters of the initilizer are correct' do
+      let(:user_identity) { '1234.5678.910' }
+      let(:uid) { user_identity }
+
+      it 'returns true' do
+        expect(user_identity_service.valid?).to be true
```
Comment (mrubi-rootstrap) [inline spec/user_identity_spec.rb:50]: I think this can be rephrased as

```ruby
expect(user_identity_service).to be_valid
```

File: apple_sign_in.gemspec [hunk 7]
```diff
@@ -1,34 +1,41 @@
...
+  spec.metadata['homepage_uri'] = spec.homepage
+  spec.metadata['source_code_uri'] = 'https://github.com/rootstrap/apple_sign_in'
+  spec.metadata['changelog_uri'] = 'https://github.com/rootstrap/apple_sign_in'
 
   # Specify which files should be added to the gem when it is released.
   # The `git ls-files -z` loads the files in the RubyGem that have been added into git.
-  spec.files = Dir['LICENSE.txt', 'README.md', 'lib/**/*']
+  spec.files         = Dir.chdir(File.expand_path(__dir__)) do
+    `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
+  end
```
Comment (juanmanuelramallo) [inline apple_sign_in.gemspec:24]: Why this change?

File: apple_sign_in.gemspec [hunk 8]
```diff
@@ -1,34 +1,41 @@
...
-  spec.add_development_dependency 'reek'
-  spec.add_development_dependency 'rspec'
-  spec.add_development_dependency 'rubocop'
+
+  # Production dependencies
+  spec.add_dependency 'jwt', '~> 2.2'
+
+  # Development dependencies
+  spec.add_development_dependency 'byebug', '~> 11.1'
+  spec.add_development_dependency 'railties', '~> 6.0'
```
Comment (juanmanuelramallo) [inline apple_sign_in.gemspec:34]: Why do we need this?

File: apple_sign_in.gemspec (hunk 7 above)
Comment (fedeagripa) [reply apple_sign_in.gemspec:24]: yup, I would prefer the previous one, git diff gives you some headaches time to time

File: apple_sign_in.gemspec (hunk 8 above)
Comment (fedeagripa) [reply apple_sign_in.gemspec:34]: we needed it for this : `require 'active_support/core_ext/hash'`

File: lib/apple_sign_in/user_identity.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: false
+
+module AppleSignIn
+  class UserIdentity
+    APPLE_KEY_URL = 'https://appleid.apple.com/auth/keys'.freeze
```
Comment (fedeagripa) [inline lib/apple_sign_in/user_identity.rb:5]: You don't need to freeze strings if you set `true` at the beginning of the file

File: lib/apple_sign_in/user_identity.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    end
+
+    def match_user_identity?(token_data)
+      user_identity && user_identity == token_data['sub']
+    end
+
+    def apple_key_hash
+      response = Net::HTTP.get(URI.parse(APPLE_KEY_URL))
+      certificate = JSON.parse(response)
+      ActiveSupport::HashWithIndifferentAccess.new(certificate['keys'].first)
```
Comment (vitogit) [inline lib/apple_sign_in/user_identity.rb:37]: I think that you need to specify the `alg` value to get the correct key from the certificate. Something like this :
```
    header = JSON.parse(Base64.decode64(jwt.split('.').first))
    alg = header['alg']
    kid = header['kid']

    matching_key = certificate['keys'].select{ |h| h['kid'] == kid }
    key_hash = ActiveSupport::HashWithIndifferentAccess.new(matching_key.first)

```

Comment (TimoPeraza) [reply lib/apple_sign_in/user_identity.rb:37]: why the ```alg```? or did you mean specify the ```kid```?

Comment (vitogit) [reply lib/apple_sign_in/user_identity.rb:37]: yes, the `kid`

Comment (mrubi-rootstrap) [reply lib/apple_sign_in/user_identity.rb:37]: Yes, since Apple may have several keys it should find the one with the matching kid instead of using the first one. I don't think searching for a matching alg is necessary though, kid should be unique.
Yet I think that for starters it's fine to use the first certificate and postpone matching the key id until the definition of a specific test for that

File: spec/user_identity_spec.rb [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  let(:private_key) { OpenSSL::PKey::RSA.generate(2048) }
+  let(:jwk) { JWT::JWK.new(private_key) }
+  let(:jwt) do
+    {
+      iss: jwt_iss,
+      aud: jwt_aud,
+      exp: jwt_exp,
+      iat: jwt_iat,
+      sub: jwt_sub,
+      email: 'timothy@rootstrap.com',
```
Comment (vitogit) [inline spec/user_identity_spec.rb:20]: Not a good idea to use personal emails in open source projects. You don't know where the code will end 😆

Comment (TimoPeraza) [reply spec/user_identity_spec.rb]: But my email is in the gemspec 🤷‍♂️ and i think it's not a personal email.
